examples/custom-server/server.js

```diff
@@ -0,0 +1,33 @@
+const { createServer } = require('http')
+const { parse } = require('url')
+const next = require('next')
+const accepts = require('accepts')
+
+const app = next('.', { dev: true })
+const handle = app.getRequestHandler()
+
+app.prepare()
+.then(() => {
+  createServer((req, res) => {
+    const accept = accepts(req)
+    const type = accept.type(['json', 'html'])
+    if (type === 'json') {
+      handle(req, res)
+      return
+    }
+
+    const { pathname, query } = parse(req.url, true)
+
+    if (pathname === '/a') {
+      app.render(req, res, '/b', query)
+    } else if (pathname === '/b') {
+      app.render(req, res, '/a', query)
+    } else {
```

Comment:
@rauchg @nkzawa The `Link` implementation prevents access to the "desired" window object on the client of the decorated route during `getInitialProps`. You can test this by logging `window.location` in and out of `getInitialProps` and following a `Link`. The `window.location` will represent that of the `href` and not `as`, so things like are not accessible for data fetching. In contrast, the back button will give you the "correct" `window.location` during `getInitialProps` because it is a full render. Thoughts?